### PR TITLE
Upgrade to Node 12

### PR DESCRIPTION
**Pull request by emcsween**
### Description

Node 10 is EOL on April 30th 2021. We're moving to Node 12, which is going to be in maintenance until April 30th 2022.

#### Related Issues / PRs

overleaf/issues#3517

### Review

SandboxedModule requires us to pass some globals in Node 12 explicitly. I configured these in a global test setup file, and I took the opportunity to do some more global configuration.

To set up that test setup file, I needed to upgrade mocha and sandboxed-module. Somehow, sandboxed-module couldn't import CoffeeScript modules (such as our settings module) until I installed CoffeeScript explicitly.

`Array.sort(cmp)` takes a comparator function `cmp(a,b)` and returns:
* a negative number when a < b
* a positive number when a > b
* 0 when a = b

The algorithm behind `Array.sort()` in Node 10 gave correct results when given a comparator function that approximately followed that protocol. In Node 12, the sort algorithm can return bad results if the function doesn't follow it. I reviewed the uses of sort and found two where the comparator could be improved.

#### Potential Impact

Everything that happens when we edit a document: collaboration, document storage, history

#### Manual Testing Performed

- [x] Edit a document in two windows. Check that changes in one window appear in the other.
- [x] Look at the history
- [x] Exit the editor, reopen the document. Check that the changes are persisted.
- [x] Add comments
- [x] Add tracked changes

### CODE REVIEW

**Comment by jdleesmiller** [comment]
> where the comparator could be improved.

This seems like the scariest change here. They look right to me, but it might be worth getting a read from someone more familiar with tracked changes, e.g. @briangough or @paulojreis, to double check.

**Review by briangough — COMMENTED** [review]
The change to the sorting functions looks correct.  

The existing tests don't seem to exercise that branch of the comparisons though.   I think it would be worth having a couple of test cases that do.

However, there is a comment at the top of the file [RangesTracker.js#L17-L19](https://github.com/overleaf/document-updater/blob/ecc7498720eca3f3b2a30c11d315c9491e49576b/app/js/RangesTracker.js#L17-L19)

> // This file is shared between document-updater and web, so that the server and client share
> // an identical track changes implementation. Do not edit it directly in web or document-updater,
> // instead edit it at https://github.com/sharelatex/ranges-tracker, where it has a suite of tests

Unfortunately that repo is in coffeescript, and the web version has diverged superficially because it has different decaffeination.

If the same sorting is running in the browser in web, with the new version of v8, it may already be missorting some cases, so we may actually have divergent track-changes in docupdater and web - at least with regard to the order.

The question is whether the order matters - I don't  have a definitive answer, but I think it will be an edge case so we should just fix the sort function in both and move on.

File: test/setup.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    warn: sandbox.stub(),
+    err: sandbox.stub(),
+    error: sandbox.stub()
+  }
+}
+
+// SandboxedModule configuration
+SandboxedModule.configure({
+  requires: {
+    'logger-sharelatex': stubs.logger
```
Comment (thomas-) [inline test/setup.js:23]: No more having to define the logger stubs in individual tests? Awesome news :)

File: app/js/RangesTracker.js [hunk 2]
```diff
@@ -728,8 +730,10 @@ const load = function () {
           return result
         } else if (a.i != null && b.d != null) {
           return 1
-        } else {
+        } else if (a.d != null && b.i != null) {
           return -1
+        } else {
+          return 0
```
Comment (thomas-) [inline app/js/RangesTracker.js:736]: Logic seems correct to me (and necessary due to the change to Array.sort I imagine...)

Did the tests fail before making this change out of curiosity? I'm just wondering how easy it might be for someone to accidentally break something, or if it would be helpful to have a small comment to say that it must return a number as a warning to future developers.

File: app/js/RangesTracker.js [hunk 3]
```diff
@@ -706,8 +706,10 @@ const load = function () {
           return result
         } else if (c1.op.i != null && c2.op.d != null) {
           return 1
-        } else {
+        } else if (c1.op.d != null && c2.op.i != null) {
           return -1
+        } else {
+          return 0
```
Comment (thomas-) [inline app/js/RangesTracker.js:712]: (See L733)

**Review by thomas- — COMMENTED** [review]
Looks good to me, just a small thing re: the sort comparator I'm curious about.

**Comment by emcsween** [comment]
> // This file is shared between document-updater and web, so that the server and client share
> // an identical track changes implementation. Do not edit it directly in web or document-updater,
> // instead edit it at https://github.com/sharelatex/ranges-tracker, where it has a suite of tests

Thanks. I missed that. I guess that's an indication that it's easy to miss. I'll see what we can do about that. Two options:

* make the ranges-tracker repo a proper library that gets imported in web and doc updater
* copy the tests over from the ranges-tracker repo to the web an doc updater repos

@paulojreis, WDYT?

File: app/js/RangesTracker.js (hunk 2 above)
Comment (emcsween) [reply app/js/RangesTracker.js:736]: The tests passed before. And I didn't find a case where the sort function misbehaved. It's possible that the comparator never caused problems. It's still not well defined. You can build two objects, `a` and `b`, such that `cmp(a,b) == -1` and `cmp(b,a) == -1`, which would mean that both `a < b` and `b < a`.

So, it's not a necessary change triggered by an actual problem, but it's a logic problem that has the potential to trigger undefined behaviour.

**Comment by emcsween** [comment]
I ran the tests for the modified code in RangesTracker.js in the ranges-tracker repo (see https://github.com/overleaf/ranges-tracker/pull/2). The tests continue to pass.

**Review by briangough — APPROVED** [review]

**Review by thomas- — APPROVED** [review]
